README.md
buffered-writer
===============

_Node.js project_

#### Writes buffered data ####

Version: 0.2.3

Node.js streams are not buffered, that is, when you write data to them you're doing multiple I/O calls. This module buffers the data that has to be written to disk and eases the buffer manipulation when you need to write data from different nature (strings, numbers, arrays or buffers).

#### Installation ####

```
npm install buffered-writer
```

#### Example ####

```javascript
var bw = require ("buffered-writer");

bw.open ("file")
	.on ("error", function (error){
		console.log (error);
	})
	.write ([0x00, 0x01, 0x02]) //Writes: 0x00, 0x01, 0x02
	.write (new Buffer ([0x03, 0x04]), 1, 1) //Writes: 0x04
	.write (0x0506) //Writes: 0x05, 0x06
	.write ("↑a", 1) //Writes: a (0x61)
	.close ();
```

#### Methods ####

When you call to `open()` a `Writer` instance is returned. This object inherits from `EventEmitter` and wraps a `WriteStream`. Only an `error` event is emitted. When this ocurrs, the `Writer` is closed automatically, you don't need to close it explicitly (if you try to do so you'll get another error, you cannot close twice).

- [bw.open(file[, settings])](#open)
- [Writer#close([callback])](#close)
- [Writer#line()](#line)
- [Writer#write(data[, offset[, length]])](#write)
- [Writer#writeln(data[, offset[, length]])](#writeln)

<a name="open"></a>
__bw.open(file[, settings])__  
Creates a `Writer` and opens a stream to the given file. If the file doesn't exist it is created. The possible settings are:

- bufferSize. _Number_. The buffer size. It must be greater than 0. Default is 16KB.
- encoding. _String_. The file encoding. Is only used when strings are written, that is, this parameter is ignored when arrays, numbers and buffers are written. Default is "utf8". Possible values are: "ascii", "utf8", "utf16le", "ucs2", "base64", "hex".
- append. _Boolean_. If true appends the data, if false the file is truncated. Default is false.
- mode. _Number_ | _String_. The file attributes. Default is 0644.
- start. _Number_. The offset to start writing data. Default is 0.

Returns the new `Writer` instance.

<a name="close"></a>
__Writer#close([callback])__  
Flushes the remaining data and closes the `Writer`. The callback doesn't receive any parameter.

<a name="flush"></a>
__Writer#flush(callback)__  
Flushes the current data. After the callback is executed it's safe to read back the written data, the data has been stored successfully. The callback doesn't receive any parameter.

This function is typically used when you need to ensure that the data is flushed to the disk at some point before closing the stream.

<a name="line"></a>
__Writer#line()__  
Writes a line, OS dependent, `\r\n` on Windows, `\n` otherwise.

Returns the `Writer` instance to allow method chaining.

<a name="write"></a>
__Writer#write(data[, offset[, length]])__  
Writes data. It's not flushed to the underlying I/O layer, it's queued to a buffer and when it's full it's flushed to the disk.

The data can be a Number, String, Array or Buffer.
- Number. Is stored in big endian, e.g.: 0x0102  will be stored as [0x01, 0x02], 0x01 in a lower address.
- String. The `open()`'s encoding parameter is used.
- Array. An array of numbers. They are stored in the same order.
- Buffer. The Node.js built-in Buffer.

The offset and length specifies the slice of the given piece of data that will be written. By default offset is 0 and length is the last byte/character. The offset and length are applied to the string characters, not the bytes, for example, `write("↑a", 1)` will write `a` because offset is 1. Take into account that `↑` is encoded with 3 bytes in utf8.

Returns the `Writer` instance to allow method chaining.

<a name="writeln"></a>
__Writer#writeln(data[, offset[, length]])__  
Does the same as `write()` but stringifies the data and concatenates an EOL, OS dependent, `\r\n` on Windows, `\n` otherwise.

CHANGES
v0.2.3 (11 Dec 2012)
	Replaced "errno-codes" dependency by "error-provider".

v0.2.2 (11 Dec 2012)
	Added "mode" and "start" to the "open()" parameters.
	Added "writeln()".
	Added "flush()".
	Renamed "newLine()" by "line()".

v0.2.1 (09 Dec 2012)
	An INVALID_DATA error is emitted if the data to write is invalid.

v0.2.0 (09 Dec 2012)
	Complete code revision and refactor.
	Removed "touch()".
	Added "open()".
	Minor improvements.

v0.1.12 (09 Oct 2012)
	Minor bugfixes.

v0.1.11 (17 Aug 2012)
	Updated errno-codes module.

v0.1.10 (05 Aug 2012)
	Better error management.

v0.1.9 (24 Jul 2012)
	Added touch() function.

v0.1.8 (22 Jul 2012)
	Improved code when the data to write had a length of 0.

v0.1.7 (24 May 2012)
	Fixed "error" event. Now is correctly emitted.

v0.1.6 (04 May 2012)
	Wow... Fixed noob error.

v0.1.5 (04 May 2012)
	Fixed bug when the data length exceeds the buffer size.
	Fixed bug with utf8 encoding.

v0.1.4 (02 May 2012)
	Now the constructor receives an object in literal notation for the settings.
	The "close()" function now receives a callback for being notified when the
		data has been flushed to the kernel and it's safe to read it back.

v0.1.3 (30 Apr 2012)
	Improved number to hex array function.

v0.1.2 (29 Apr 2012)
	Now it's possible to write more than one byte passing a number, eg.:
		"write(0x1234)" will write 0x12, 0x34.

v0.1.1 (28 Apr 2012)
	Fixed parameters in "write()" function.

v0.1.0 (28 Apr 2012)
	No more callbacks. The buffer is flushed using a WriteStream.

v0.0.1 (28 Apr 2012)
	First commit.

tests/run.js
"use strict";

var Runner = require ("mocha-runner");

new Runner ({
	tests: ["buffered-writer.js"]
}).run (function (error){
	if (error) console.log (error);
});

test/playground.js
var BufferedWriter = require ("../build/buffered-writer");

test/tests/utf8.js
var BufferedWriter = require ("../../build/buffered-writer");

new BufferedWriter ("tmp", { encoding: "utf8", bufferSize: 2 }).write ("↑↓←→").close ();

buffered-writer.js
"use strict";

var EVENTS = require ("events");
var FS = require ("fs");
var ep = require ("error-provider");

ep.create (ep.next (), "INVALID_BUFFER_SIZE", 
		"The buffer size must be greater than 0.");
ep.create (ep.next (), "STREAM_CLOSED", 
		"The stream is already closed, cannot write nor close it again.");
ep.create (ep.next (), "INVALID_OFFSET_LENGTH", 
		"The offset or length parameters are not valid (offset={offset}, " +
		"length={length}, length-offset<0).");
ep.create (ep.next (), "INVALID_DATA", 
		"The data can only be a Number, String, Array or Buffer");

var BUFFER_SIZE = 16384;
var EOL = process.platform === "win32"
		? new Buffer ([0x0d, 0x0a])
		: new Buffer ([0x0a]);
var strEOL = EOL.toString ();

var bw = module.exports = {};

bw.open = function (file, args){
	args = args || {};
	
	if (args.bufferSize === 0) args.bufferSize = -1;
	args.bufferSize = args.bufferSize || BUFFER_SIZE;
	if (args.bufferSize < 1) throw ep.get ("INVALID_BUFFER_SIZE");
	
	return new Writer (file, args);
};

var Writer = function (file, args){
	EVENTS.EventEmitter.call (this);
	
	this._bufferSize = args.bufferSize;
	this._encoding = args.encoding;
	this._buffer = null;
	this._offset = 0;
	this._closed = false;
	
	var me = this;
	this._stream = FS.createWriteStream (file, {
		flags: args.append ? "a" : "w",
		encoding: args.encoding,
		mode: args.mode,
		start: args.start
	}).on ("error", function (error){
		me.emit ("error", error);
	}).on ("drain", function (){
		if (me._flushCallback){
			me._flushCallback ();
			me._flushCallback = null;
		}
	});
};

Writer.prototype = Object.create (EVENTS.EventEmitter.prototype);
Writer.prototype.constructor = Writer;

Writer.prototype._error = function (error){
	if (this.listeners ("error").length !== 0){
		this.emit ("error", error);
	}else{
		throw error;
	}
};

Writer.prototype._flush = function (){
	var written = this._stream.write (
			new Buffer (this._offset !== this._bufferSize
					? this._buffer.slice (0, this._offset)
					: this._buffer));
	this._offset = 0;
	return written;
};

Writer.prototype._write = function (data, offset, length){
	var me = this;

	if (!this._buffer){
		this._buffer = new Buffer (this._bufferSize);
	}
	
	var bytes = length + this._offset > this._bufferSize ?
			this._bufferSize - this._offset :
			length;
	
	data.copy (this._buffer, this._offset, offset, offset + bytes);
	this._offset += bytes;
	offset += bytes;
	length -= bytes;
	if (this._offset === this._bufferSize){
		this._flush ();
		if (length !== 0){
			this._write (data, offset, length);
		}
	}
};

Writer.prototype.close = function (cb){
	if (this._closed){
		return this._error (ep.get ("STREAM_CLOSED"));
	}

	if (this._offset){
		this._flush ();
	}
	
	var me = this;
	this._stream.on ("close", function (){
		if (cb) cb ();
	});
	this._stream.end ();
	this._closed = true;
};

Writer.prototype.flush = function (cb){
	if (this._offset === 0) return cb ();
	if (!this._flush ()){
		this._flushCallback = cb;
	}else{
		cb ();
	}
};

Writer.prototype.line = function (){
	if (this._closed){
		return this._error (ep.get ("STREAM_CLOSED"));
	}
	this._write (EOL, 0, EOL.length);
	return this;
};

var toHexArray = function (n){
	var array = [];
	do{
		array.unshift (n & 0xff);
		n = n >>> 8;
	}while (n);
	return array;
};

Writer.prototype.write = function (buffer, offset, length){
	if (this._closed){
		return this._error (ep.get ("STREAM_CLOSED"));
	}
	
	offset = offset || 0;
	var stringError;
	
	var type = typeof buffer;
	if (type === "number"){
		buffer = toHexArray (buffer);
		length = length || buffer.length - offset;
		buffer = new Buffer (buffer);
	}else if (type === "string"){
		stringError = length - offset < 0;
		buffer = buffer.substr (offset, length);
		offset = 0;
		length = Buffer.byteLength (buffer, this._encoding);
		buffer = new Buffer (buffer, this._encoding);
	}else if (Array.isArray (buffer)){
		buffer = new Buffer (buffer);
		length = length || buffer.length - offset;
	}else if (Buffer.isBuffer (buffer)){
		length = length || buffer.length - offset;
	}else{
		var me = this;
		this.close (function (){
			me._error (ep.get ("INVALID_DATA"));
		});
		return;
	}
	
	if (stringError || length < 0){
		var me = this;
		this.close (function (){
			me._error (ep.get ("INVALID_OFFSET_LENGTH", {
				offset: offset,
				length: length
			}));
		});
		return;
	}
	
	if (length !== 0){
		this._write (buffer, offset, length);
	}
	return this;
};

Writer.prototype.writeln = function (buffer, offset, length){
	return this.write (buffer + strEOL, offset, length);
};

examples/writeText.js
var BufferedWriter = require ("../build/buffered-writer");

//The BufferedWriter truncates the file because append == false
new BufferedWriter ("file", { encoding: "utf8" })
	.on ("error", function (error){
		console.log (error);
	})
	
	//From the beginning of the file:
	.write ("First line") //Writes "First line"
	.newLine () //Writes EOL (OS dependent; \r\n on Windows, otherwise \n)
	.write ("Second line") //Writes "Second line"
	.close (); //Closes the writer. A flush is implicitly done.

//The BufferedWriter appends content to the end of the file because append == true
new BufferedWriter ("file", { encoding: "utf8", append: true })
	.on ("error", function (error){
		console.log (error);
	})
	
	//From the end of the file:
	.newLine () //Writes EOL (OS dependent; \r\n on Windows, otherwise \n)
	.write ("Third line") //Writes "Third line"
	.close (); //Closes the writer. A flush is implicitly done.
	
//The file contains:
/*
	First line
	Second line
	Third line
*/

examples/writeBytes.js
var BufferedWriter = require ("../build/buffered-writer");

//The BufferedWriter truncates the file because append == false
new BufferedWriter ("file")
	.on ("error", function (error){
		console.log (error);
	})
	
	//From the beginning of the file:
	.write ([0x00, 0x01, 0x02], 0, 3) //Writes 0x00, 0x01, 0x02
	.write (new Buffer ([0x03, 0x04]), 1, 1) //Writes 0x04
	.write (0x0506) //Writes 0x05, 0x06
	.close (); //Closes the writer. A flush is implicitly done.

//The BufferedWriter appends content to the end of the file because append == true
new BufferedWriter ("file", { append: true })
	.on ("error", function (error){
		console.log (error);
	})
	
	//From the end of the file:
	.write (0xFF) //Writes 0xFF
	.close (); //Closes the writer. A flush is implicitly done.

//The file contains: 0x00, 0x01, 0x02, 0x04, 0x05, 0x06, 0xFF